qiskit_experiments/curve_analysis/models.py

```diff
@@ -0,0 +1,354 @@
+                    "Now '.fit_func' is a string representation of model. "
+                    "This warning will be removed and support for callable is dropped "
+                    "in Qiskit Experiments 0.5.",
+                    DeprecationWarning,
+                    stacklevel=3,
+                )
+                # Use callable as-is with parsing function signature
+                expr_callable = expr
+                signature = list(inspect.signature(expr_callable).parameters.keys())
+                params = signature[1:]
+            else:
+                # Generate function from string
+                expr_callable, params = _compile_expr(expr)
+            curve_models.append(expr_callable)
+            curve_params.append(params)
+
+        def _composite(x, allocation, **kwargs):
+            y = np.zeros(x.shape)
+            for f_idx, (func, params) in enumerate(zip(curve_models, curve_params)):
+                locs = allocation == f_idx
+                y[locs] = func(x=x[locs], **{p: kwargs[p] for p in params})
+            return y
+
+        unite_params = []
+        for params in curve_params:
+            unite_params.extend(p for p in params if p not in unite_params)
+
+        return _composite, unite_params
+
+    @property
+    def definitions(self) -> Iterator[SeriesDef]:
+        """Return series definitions."""
+        yield from self._series_defs
+
+    def eval_with_uncertainties(
+        self,
+        x: np.ndarray,
+        params: Dict[str, uncertainties.UFloat],
+        model_index: int,
+    ) -> np.ndarray:
+        """Compute Y values with error propagation.
+
+        Args:
+            x: X values.
+            params: Fitter parameters with uncertainties.
+            model_index: Index of model to compute.
+
+        Returns:
+            Y values with uncertainty (uarray).
+        """
+        model_expr = self._series_defs[model_index].fit_func
+
+        # Create function that performs error propagation
+        func, this_params = _compile_expr(model_expr)
+        sub_params = {pname: params[pname] for pname in this_params}
+        wrapfunc = np.vectorize(uncertainties.wrap(func))
+
+        return wrapfunc(x=x, **sub_params)
+
+    def _parse_params(self):
+        self.independent_vars = ["x", "allocation"]
+        self._param_names = self._param_root_names
+        self._func_haskeywords = True
+        self._func_allargs = self.independent_vars + self.param_names
+        self.def_vals = {}
+
+    def __getstate__(self):
+        # Local function cannot be pickled
+        state = self.__dict__.copy()
+        del state["func"]
+
+        return state
+
+    def __setstate__(self, state):
+        self.__dict__.update(state)
+
+        # Initialize non-pickled callable from string expressions
+        composite_func, _ = self._compose_functions(
+            exprs=[series_def.fit_func for series_def in state["_series_defs"]]
+        )
+        self.func = composite_func
```

Comment:
This is for pickle (though it's removed now). Because pickle serializes everything stored in the instance, i.e. not only constructor args but also all instance members including private variable of python callable that cannot be serialized without `dill`. Experiments supports two paths for serialization, i.e. JSON and pickle.